File: Python/DataStructures/linked_list.py
"""LinkedList Data Structure"""


class LinkedList:
    def __init__(self):
        self.head = None
        self.length = 0

    def insert_first(self, data):
        """Inserts a node with the given data at the beginning of the list."""
        new_node = LLNode(data)

        if self.head is None:
            self.head = new_node
        else:
            new_node.next = self.head
            self.head = new_node

        self.length = self.length + 1

    def insert_last(self, data):
        """Inserts a node with the given data at the end of the list."""
        new_node = LLNode(data)

        if self.head is None:
            self.head = new_node
            self.length = self.length + 1
            return

        current = self.head
        # get reference to the last node
        while current.next is not None:
            current = current.next

        current.next = new_node

        self.length = self.length + 1

    def insert(self, data, index):
        """Inserts a node with the given data at the specified index position in the list.
        If the given index is out of bounds, it returns false indicating the operation was
        not successful. Otherwise returns true."""

        # if index is within bounds (i.e., index exists)
        if 0 <= index <= self.length - 1:
            new_node = LLNode(data)

            if index == 0:
                new_node.next = self.head
                self.head = new_node
                self.length = self.length + 1
                return True

            current = self.head
            curr_idx = 0  # current node's index

            # locate the node before the node at the index
            while current.next is not None:
                # if the current node is the node before the node at the index
                if curr_idx == index - 1:
                    break
                current = current.next
                curr_idx = curr_idx + 1

            node_at_index = current.next
            new_node.next = node_at_index
            current.next = new_node
            self.length = self.length + 1
            return True

        return False

    def remove_first(self):
        """Removes the first node from the list and returns its value. Returns None if list is empty."""
        if self.head is None:
            return None

        first = self.head
        first_data = first.data
        self.head = first.next
        self.length = self.length - 1

        return first_data

    def remove_last(self):
        """Removes the last node from the list and returns its value. If the list is empty, returns None."""
        if self.head is None:
            return None

        if self.head.next is None:
            last_data = self.head.data
            self.head = self.head.next
            return last_data

        current = self.head

        # get reference to the node before last
        while current.next is not None and current.next.next is not None:
            current = current.next

        last = current.next
        last_data = last.data
        current.next = None
        self.length = self.length - 1

        return last_data

    def remove(self, index):
        """Removes a node at the specified index. Returns None if the index is out of bounds,
        otherwise returns the value at the removed node."""

        # if index is within bounds (i.e, it exists)
        if 0 <= index <= self.length - 1:
            if index == 0:
                the_node = self.head  # node to be removed
                self.head = the_node.next
                the_node.next = None  # detach the node from the rest of the list
                self.length = self.length - 1
                return the_node.data

            current = self.head
            curr_idx = 0  # index of the current node

            # scan the list to the node before the node to be removed
            while curr_idx < index - 1:
                current = current.next
                curr_idx = curr_idx + 1

            the_node = current.next  # node to be removed
            node_after = the_node.next  # node after the node to be removed
            current.next = node_after
            the_node.next = None
            self.length = self.length - 1
            return the_node.data

        return None

    def remove_by_data(self, data):
        """Removes the first node with the given value. Returns the value
        if found, none otherwise."""

        if self.head is None:
            return None

        if self.head.data == data:
            the_node = self.head  # the node to be removed
            self.head = the_node.next
            the_node.next = None  # detach the node from the rest of the list
            self.length = self.length - 1
            return data

        current = self.head

        while current.next is not None and current.next.data != data:
            current = current.next

        # if the node is found
        if current.next is not None:
            the_node = current.next
            node_after = the_node.next  # node after the node to be removed
            current.next = node_after
            the_node.next = None
            self.length = self.length - 1
            return data

        return None

    def get(self, index):
        """Retrieves the value of the node at the specified index. Returns None if index does not exist."""

        # if index exists (also implies that the list is not empty)
        if 0 <= index <= self.length - 1:
            current = self.head
            curr_idx = 0

            while current is not None:
                if curr_idx == index:
                    return current.data
                current = current.next
                curr_idx = curr_idx + 1

        return None

    def size(self):
        """Returns the length of the list."""
        return self.length

    def is_empty(self):
        return self.length == 0

    def get_position(self, data):
        """Returns the position of the first node with the given value, -1 if not found."""
        current = self.head
        curr_idx = 0

        while current is not None:
            if current.data == data:
                return curr_idx
            current = current.next
            curr_idx = curr_idx + 1

        return -1

    def clear(self):
        """Empties the list."""
        self.head = None
        self.length = 0


class LLNode:
    def __init__(self, data):
        self.data = data
        self.next = None

File: JavaScript/DataStructures/linked_list.js
/**
 * Linked List implementation using ES5 constructor function. This implementation
 * uses private variables to implement the head and length of the list in order to prevent
 * direct access to the head node and length through the dot operator as is the case
 * with ES6 class.
 */
export function LinkedList() {
  // private variables
  let head = null;
  let length = 0;

  /**
   * Gets the head element node.
   * @returns the head element
   */
  this.getHead = function () {
    return head;
  };

  /**
   * Adds a node of the given data to the beginning of the list.
   * @param {*} data data of node to be added
   */
  this.addFirst = function (data) {
    let newNode = new LLNode(data);
    // if list is empty
    if (head === null) {
      head = newNode;
    } else {
      newNode.setNext(head);
      head = newNode;
    }

    ++length;
  };

  /**
   * Adds a node of the given data to the end of the list.
   * @param {*} data data of node to be added
   */
  this.addLast = (data) => {
    let newNode = new LLNode(data);
    // if list is empty
    if (head === null) {
      head = newNode;
    } else {
      let current = head;

      while (current.getNext() !== null) {
        current = current.getNext();
      }

      current.setNext(newNode);
    }

    ++length;
  };

  /**
   * Adds a node of the given data at the specified index in the list. Throws an
   * error if index is out of bounds.
   * @param {*} data data of node to be added
   * @param {*} index position at which to add the node
   */
  this.add = function (data, index) {
    let newNode = new LLNode(data);
    // if index exist (i.e., index is within bounds)
    if (index >= 0 && index <= length - 1) {
      // if adding at the first position
      if (index === 0) {
        newNode.setNext(head);
        head = newNode;
        length++;
        return;
      }

      let current = head; // current node
      let currIdx = 0; // current index pointer
      // scan the list to the node before the index
      while (current.getNext() !== null) {
        // if the current node is the node before the index
        if (currIdx === index - 1) {
          break;
        }
        current = current.getNext();
        currIdx++;
      }

      let nodeAtIndex = current.getNext();
      newNode.setNext(nodeAtIndex);
      current.setNext(newNode);
      length++;
      return;
    }

    throw Error('Index out of bounds!');
  };

  /**
   * Removes the first node in the list.
   * @returns the data at the first node removed or null if list is empty.
   */
  this.removeFirst = function () {
    // if list is empty
    if (head === null) {
      return null;
    }

    let firstdata = head.getData();
    head = head.getNext();
    length--;

    return firstdata;
  };

  /**
   * Removes the last node in the list.
   * @returns data at the last node removed or null if list is empty
   */
  this.removeLast = function () {
    // if list is empty
    if (head === null) {
      return null;
    }

    let lastdata;

    // if the list has only one node
    if (head.getNext() === null) {
      lastdata = head.getData();
      head = null;
      length--;
      return lastdata;
    }

    let current = head;
    // scan the list to a node before the last
    while (current.getNext().getNext() !== null) {
      current = current.getNext();
    }

    lastdata = current.getNext().getData();
    current.setNext(null); // set the last node to null;
    length--;

    return lastdata;
  };

  /**
   * Removes the data of node at the given index and returns it.
   * @param {*} index position of node to remove
   * @returns data at the removed node or null if index is invalid
   */
  this.remove = (index) => {
    // if index exists (i.e., within bounds)
    if (index >= 0 && index <= length - 1) {
      let removeddata;
      if (index === 0) {
        removeddata = head.getData();
        head = head.getNext();
        length--;
        return removeddata;
      }

      let current = head;
      let currIdx = 0;
      // scan the list to a node before the desired node to remove
      while (currIdx < index - 1) {
        current = current.getNext();
        currIdx++;
      }

      let nodeToRemove = current.getNext();
      removeddata = nodeToRemove.getData();
      let nodeAfter = nodeToRemove.getNext();
      current.setNext(nodeAfter);
      length--;

      return removeddata;
    }
  };

  /**
   * Returns the data of the node at the given index position.
   * @param {*} index position of node whose data is to be retrieved
   * @returns data of the node at the given index or undefined if index is invalid
   */
  this.dataAt = (index) => {
    // if index is valid
    if (index >= 0 && index <= length - 1) {
      let current = head;
      let currIdx = 0;

      while (currIdx < index) {
        current = current.getNext();
        currIdx++;
      }

      return current.getData();
    }
  };

  /**
   * Returns the index of the first node whose data equals the given data.
   * @param {*} data data of the node whose index should be retrieved
   * @returns index position of the node with the given data or -1 if index is out of
   * bounds
   */
  this.indexOf = (data) => {
    let current = head;
    let currIdx = 0;
    // scan the list to find the node with the given data
    while (current !== null && current.getData() !== data) {
      current = current.getNext();
      currIdx++;
    }

    if (current !== null) {
      return currIdx;
    }

    return -1;
  };

  /**
   * Returns the size of the list.
   * @returns the size of the list
   */
  this.size = () => {
    return length;
  };

  /**
   * Empties the list.
   */
  this.clear = () => {
    head = null;
    length = 0;
  };
}

/**
 * Linked List node implementation using ES5 constructor function.
 * @param {*} data data at the node to be created
 */
export function LLNode(data) {
  // if data is truthy
  if (data === undefined || data === null) {
    throw Error('data cannot be null or undefined');
  }

  // private variables
  let _data = data;
  let _next = null;

  this.getData = function () {
    return data;
  };

  this.setdata = function (data) {
    if (data !== undefined) {
      _data = data;
    }
  };

  this.getNext = function () {
    return _next;
  };

  this.setNext = function (next) {
    if (next === null || next instanceof LLNode) {
      _next = next;
    } else {
      throw Error('Attempt to add an invalid node object to this list');
    }
  };
}
